go/tools/gazelle/merger/merger.go

```diff
@@ -46,112 +46,309 @@ var (
 	}
 )
 
-// MergeWithExisting merges newfile with an existing build file at
-// existingFilePath and returns newfile.
-func MergeWithExisting(newfile *bzl.File, existingFilePath string) (*bzl.File, error) {
-	b, err := ioutil.ReadFile(existingFilePath)
+// MergeWithExisting merges genfile with an existing build file at
+// existingFilePath and returns genfile.
+func MergeWithExisting(genFile *bzl.File, existingFilePath string) (*bzl.File, error) {
+	oldData, err := ioutil.ReadFile(existingFilePath)
 	if err != nil {
 		return nil, err
 	}
-	f, err := bzl.Parse(newfile.Path, b)
+	oldFile, err := bzl.Parse(genFile.Path, oldData)
 	if err != nil {
 		return nil, err
 	}
-	if err := shouldIgnore(f); err != nil {
+	if err := shouldIgnore(oldFile); err != nil {
 		return nil, err
 	}
 
+	oldStmt := oldFile.Stmt
 	var newStmt []bzl.Expr
-	for _, s := range newfile.Stmt {
-		c, ok := s.(*bzl.CallExpr)
+	for _, s := range genFile.Stmt {
+		genRule, ok := s.(*bzl.CallExpr)
 		if !ok {
-			return nil, fmt.Errorf("got %v expected only CallExpr in %q", s, newfile.Path)
+			return nil, fmt.Errorf("got %v expected only CallExpr in %q", s, genFile.Path)
 		}
-		other, err := match(f, c)
-		if err != nil {
-			return nil, err
-		}
-		if other == nil {
-			newStmt = append(newStmt, c)
+		i, oldRule := match(oldFile, genRule)
+		if oldRule == nil {
+			newStmt = append(newStmt, genRule)
 			continue
 		}
-		if name(c) == "load" {
-			mergeLoad(c, other, f)
+
+		var mergedRule bzl.Expr
+		if kind(oldRule) == "load" {
+			mergedRule = mergeLoad(genRule, oldRule, oldFile)
 		} else {
-			merge(c, other)
+			mergedRule = mergeRule(genRule, oldRule)
 		}
+		oldStmt[i] = mergedRule
 	}
-	f.Stmt = append(f.Stmt, newStmt...)
-	return f, nil
+
+	oldFile.Stmt = append(oldStmt, newStmt...)
+	return oldFile, nil
 }
 
-// shouldIgnore checks whether "gazelle:ignore" appears at the beginning of
-// a comment before or after any top-level statement in the file.
-func shouldIgnore(oldFile *bzl.File) error {
-	for _, s := range oldFile.Stmt {
-		for _, c := range s.Comment().After {
-			if strings.HasPrefix(c.Token, gazelleIgnore) {
-				return GazelleIgnoreError{c.Start}
+// merge combines information from gen and old and returns an updated rule.
+// Both rules must be non-nil and must have the same kind and same name.
+func mergeRule(gen, old *bzl.CallExpr) *bzl.CallExpr {
+	genRule := bzl.Rule{Call: gen}
+	oldRule := bzl.Rule{Call: old}
+	merged := *old
+	merged.List = nil
+	mergedRule := bzl.Rule{Call: &merged}
+
+	for _, k := range oldRule.AttrKeys() {
+		oldAttr := oldRule.Attr(k)
+		var mergedAttr bzl.Expr
+		if !mergeableFields[k] {
+			mergedAttr = oldAttr
```

Comment:
long else blocks are sad...  
oldAttr is known to exists, so here just
mergedRule.SetAttr(k, oldAttr)
continue

and then avoid the long else as the following code requires no indent.